compiler/src/middle_end/optimize_tail_calls.re

```diff
@@ -1,492 +1,26 @@
-/**
-This module transforms tail-recursive directly and mutually recursive functions into loops.
-
-Until WebAssembly supports tail calls, this module aims to transform this program:
-```
-let rec isEven = (x) => {
-  if (x <= 1) {
-    x == 0
-  } else {
-    isOdd(x - 1)
-  }
-}
-and isOdd = (y) => {
-  if (y <= 1) {
-    y == 1
-  } else {
-    isEven(y - 1)
-  }
-}
-```
-
-Into this program:
-```
-let continue = box 0,
-next = box 0,
-_x = box 0,
-_y = box 0;
-
-let rec isEven = (x) => {
-  _x := x;
-  continue := true;
-  next := _isEven;
-  let returnValue = box 0;
-  while (unbox continue) {
-    continue := false;
-    returnValue := (unbox next)()
-  };
-  unbox returnValue
-}
-and isOdd = (y) => {
-  _y := y;
-  continue := true;
-  next := _isOdd;
-  let returnValue = box 0;
-  while (unbox continue) {
-    continue := false;
-    returnValue := (unbox next)()
-  };
-  unbox returnValue
-}
-and _isEven = () => {
-  let x = unbox _x;
-  if (x <= 1) {
-    x == 0
-  } else {
-    _y := x - 1;
-    continue := true;
-    next := _isOdd
-  }
-}
-and _isOdd = () => {
-  let y = unbox _y;
-  if (y <= 1) {
-    y == 0
-  } else {
-    _x := y - 1;
-    continue := true;
-    next := _isEven
-  }
-}
-```
-
-Things to note:
-  * Two references, `continue` and `next` are available to the entire set of tail-recursive funtions.
-  * References are created for all arguments to the functions. These are available to the entire set of recursive
-    functions.
-  * Each function is transformed into two functions-- an invocation function and an iteree function. The invocation
-    function is called when the function is called from an outside source, and controls the loop that calls the iteree.
-    The iteree function is a thunk that first binds the would-be arguments from the references to names, then proceeds
-    with the original function body with minor alterations, the most important being a replacement of tail calls with a
-    setup to continue the iteration.
-  * Further optimizations (such as using only a single function) could be done for directly-recursive functions, but
-    these are not implemented at this time.
-*/;
-
 open Anftree;
```

Comment:
So this file still needs to exist?

Reply:
Yeah. We still need to tell the compiler when we need to do a tail call.